### PR TITLE
Implement design

### PR DESCRIPTION
**Pull request by hug-dev**
Implement the design of the `psa-crypto` crate as discussed in #8 
This is not finished but the PR is done for early reviews/comments.

I decided to remove `id` from the `Attributes` as it seemed weird to have this value as both an input of the function (for persistant keys keys) and an output. I replaced it with `Option<u32>` which represent a tentative of ID for the called that wants to generate a persistent key.

Most of the `Attributes` and `Algorithm` structures were taken from the Parsec Rust Interface. Because `ANY_HASH` can only be used for hash-and-sign algorithm, I decided to remove it from the `Hash` enumeration and to create the `SignHash` one, just for asymmetric signature functions.

Things yet to be done:
* add the conversion of Rust abstracted types to FFI types and the other way around
* complete documentation
* add examples
* add the interface for Secure Element drivers (possibly under a feature)

### CODE REVIEW

File: psa-crypto/src/types/key.rs [hunk 1]
```diff
@@ -0,0 +1,768 @@
...
+    pub lifetime: Lifetime,
+    /// Intrinsic category and type of the key
+    pub key_type: Type,
+    /// Size of the key in bits
+    pub bits: usize,
+    /// Policy restricting the permitted usage of the key
+    pub policy: Policy,
+}
+
+impl Attributes {
```
Comment (ionut-arm) [inline psa-crypto/src/types/key.rs:27]: Not part of this PR, but we should have a `validate(&self)` method here to check that all the fields are in a valid state (e.g. the `Type`, `Algorithm` and `bits` are in a consistent state.

File: psa-crypto/src/operations/key_management.rs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        id,
+        handle: Some(handle),
+    })
+}
+
+/// Destroy a key
+///
+/// # Safety
+///
+/// blablabla
```
[the remark this answers is not in the document]
Comment (hug-dev) [reply psa-crypto/src/operations/key_management.rs:37]: This is not a view on what I think about safety 😄

File: psa-crypto/src/types/key.rs (hunk 1 above)
Comment (hug-dev) [reply psa-crypto/src/types/key.rs:27]: Would be good indeed!
We can also think about making some of those fields (all) private and creating specific constructors or builders that only build consistent attributes.

File: psa-crypto/src/types/key.rs [hunk 3]
```diff
@@ -0,0 +1,768 @@
...
+        })
+    }
+
+    pub(crate) fn close_handle(self, handle: psa_crypto_sys::psa_key_handle_t) -> Result<()> {
+        if self.handle.is_none() {
+            status_to_result(unsafe { psa_crypto_sys::psa_close_key(handle) })
+        } else {
+            Ok(())
+        }
+    }
```
[the remark this answers is not in the document]
Comment (hug-dev) [reply psa-crypto/src/types/key.rs:410]: Oh yes you are right, it would make more sense to directly use the one in `self`. I think I went through several refactoring of this so maybe I mixed things up.

File: psa-crypto/src/types/key.rs [hunk 4]
```diff
@@ -0,0 +1,768 @@
...
+    /// Create a new Id from a persistent key ID
+    pub fn from_persistent_key_id(id: u32) -> Self {
+        Id { id, handle: None }
+    }
+}
+
+impl From<Attributes> for psa_crypto_sys::psa_key_attributes_t {
+    fn from(_attributes: Attributes) -> Self {
+        unsafe { psa_crypto_sys::psa_key_attributes_init() }
+    }
```
[the remark this answers is not in the document]
Comment (hug-dev) [reply psa-crypto/src/types/key.rs:423]: Yes! This is now just a stub to make it compile.

File: psa-crypto/src/types/status.rs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+//! # PSA Status Codes
+//!
+//! This module defines success and error codes returned by any PSA function.
+
+use log::error;
+
+/// Result type returned by any PSA operation
+pub type Result<T> = core::result::Result<T, Status>;
```
[the remark this answers is not in the document]
Comment (hug-dev) [reply psa-crypto/src/types/status.rs:11]: Yes you are making a good point. I think, we can do something like:
```rust
enum Status {
    Success,
    Error(Error),
}

enum Error {
    GenericError,
    NotSupported,
    ...
}
```
That would match nicely with the layout of the status definition in the specs:
```
8.1: PSA Status codes
  8.1.2: Success codes
  8.1.3: Error codes
``` 
and then:
```rust
pub type Result<T> = core::result::Result<T, Error>;
```

File: psa-crypto/src/types/status.rs [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+impl From<psa_crypto_sys::psa_status_t> for Status {
+    fn from(status: psa_crypto_sys::psa_status_t) -> Self {
+        match status {
+            psa_crypto_sys::PSA_SUCCESS => Status::Success,
+            s => {
+                error!("{} not recognised as a valid PSA status.", s);
+                Status::GenericError
+            }
```
[the remark this answers is not in the document]
Comment (hug-dev) [reply psa-crypto/src/types/status.rs:69]: Yes 🌻

File: psa-crypto/src/types/key.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (hug-dev) [reply psa-crypto/src/types/key.rs:410]: Oh wait actually there was a reason: the `handle` value inside the `Id` structure is only for `Id` that were creating through the key creations operations. It is `None` if it was created directly from a key ID and does not have a handle assigned to it. For this case we open/close the key before/after each call.

File: psa-crypto/src/types/key.rs (hunk 1 above)
Comment (ionut-arm) [reply psa-crypto/src/types/key.rs:27]: That sounds good, we could also have static pre-defined builders for common algorithms, for example.

**Comment by hug-dev** [comment]
Don't hesitate raising issues for bugs or concerns over the design. This is still at the state of "prototype", we shall be adding more tests and examples.
